File: frontend/src/components/benchmark/content/common_neighbors/implementations/common_neighbors.js
class CommonNeighbors {
    constructor() {
        this.adjacencyList = new Map();
    }

    addEdge(u, v) {
        if (!this.adjacencyList.has(u)) {
            this.adjacencyList.set(u, new Set());
        }
        if (!this.adjacencyList.has(v)) {
            this.adjacencyList.set(v, new Set());
        }
        this.adjacencyList.get(u).add(v);
        this.adjacencyList.get(v).add(u);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    commonNeighbors(u, v) {
        if (!this.adjacencyList.has(u) || !this.adjacencyList.has(v)) {
            return new Set();
        }

        const result = new Set();
        const neighborsU = this.adjacencyList.get(u);
        const neighborsV = this.adjacencyList.get(v);

        for (const node of neighborsU) {
            if (neighborsV.has(node)) {
                result.add(node);
            }
        }
        return result;
    }

    commonNeighborsCount(u, v) {
        return this.commonNeighbors(u, v).size;
    }

    predictLink(u, v, threshold = 1) {
        return this.commonNeighborsCount(u, v) >= threshold;
    }

    scoreAllPairs() {
        const scores = [];
        const nodes = Array.from(this.adjacencyList.keys());

        for (let i = 0; i < nodes.length; i++) {
            for (let j = i + 1; j < nodes.length; j++) {
                const u = nodes[i];
                const v = nodes[j];
                const count = this.commonNeighborsCount(u, v);
                if (count > 0) {
                    scores.push({ pair: [u, v], score: count });
                }
            }
        }

        scores.sort((a, b) => b.score - a.score);
        return scores;
    }

    topKPredictions(k = 10) {
        const scores = this.scoreAllPairs();
        return scores.slice(0, k);
    }
}

module.exports = CommonNeighbors;

File: frontend/src/components/benchmark/content/common_neighbors/implementations/common_neighbors.rs
use std::collections::{HashMap, HashSet};

pub struct CommonNeighbors {
    adjacency_list: HashMap<i32, HashSet<i32>>,
}

impl CommonNeighbors {
    pub fn new() -> Self {
        CommonNeighbors {
            adjacency_list: HashMap::new(),
        }
    }

    pub fn add_edge(&mut self, u: i32, v: i32) {
        self.adjacency_list.entry(u).or_insert_with(HashSet::new).insert(v);
        self.adjacency_list.entry(v).or_insert_with(HashSet::new).insert(u);
    }

    pub fn build_from_edges(&mut self, edges: &[(i32, i32)]) {
        for &(u, v) in edges {
            self.add_edge(u, v);
        }
    }

    pub fn common_neighbors(&self, u: i32, v: i32) -> HashSet<i32> {
        if let (Some(neighbors_u), Some(neighbors_v)) = (self.adjacency_list.get(&u), self.adjacency_list.get(&v)) {
            neighbors_u.intersection(neighbors_v).cloned().collect()
        } else {
            HashSet::new()
        }
    }

    pub fn common_neighbors_count(&self, u: i32, v: i32) -> usize {
        self.common_neighbors(u, v).len()
    }

    pub fn predict_link(&self, u: i32, v: i32, threshold: usize) -> bool {
        self.common_neighbors_count(u, v) >= threshold
    }

    pub fn score_all_pairs(&self) -> Vec<((i32, i32), usize)> {
        let mut scores = Vec::new();
        let nodes: Vec<i32> = self.adjacency_list.keys().cloned().collect();

        for i in 0..nodes.len() {
            for j in (i + 1)..nodes.len() {
                let u = nodes[i];
                let v = nodes[j];
                let count = self.common_neighbors_count(u, v);
                if count > 0 {
                    scores.push(((u, v), count));
                }
            }
        }

        scores.sort_by(|a, b| b.1.cmp(&a.1));
        scores
    }

    pub fn top_k_predictions(&self, k: usize) -> Vec<((i32, i32), usize)> {
        let scores = self.score_all_pairs();
        scores.into_iter().take(k).collect()
    }
}